modules/vpc/main.tf

```diff
@@ -0,0 +1,393 @@
+# -------------------------------------------------------------------------------------
+# VIRTUAL PRIVATE CLOUD (VPC)
+# 
+# This module will create a VPC that can be used for application deployments.
+#
+# The module creates a VPC with both public and private subnets. For production
+# deployments, at least three availability zones (AZs) should be used. For non-
+# production deployments one subnet should suffice. For cost savings using one 
+# NAT gateway will work. For highly-available workloads, one NAT should be used
+# per AZ.
+#
+# The module includes the following:
+#
+# - VPC
+# - Internet Gateway
+# - NAT Gateway
+# - Public Subnets
+# - Public Subnets' Route Table
+# - Public Subnets' NACL
+# - Private Subnets
+# - Private Subnets' Route Table
+# - Private Subnets' NACL
+#
+# -------------------------------------------------------------------------------------
+
+
+# -------------------------------------------
+# SET TERRAFORM REQUIREMENTS TO RUN MODULE
+# -------------------------------------------
+
+terraform {
+  required_version = ">= 1.5.5"
+
+  required_providers {
+    aws = {
+      source  = "hashicorp/aws"
+      version = ">= 5.0"
+    }
+  }
+}
+
+
+# -------------------------------------------
+# RETRIEVE REGION INFORMATION
+# -------------------------------------------
+
+data "aws_region" "current" {}
+
+data "aws_availability_zones" "current" {
+  filter {
+    name   = "region-name"
+    values = [data.aws_region.current.name]
+  }
+
+  filter {
+    name   = "zone-type"
+    values = ["availability-zone"]
+  }
+
+  lifecycle {
+    # If the number of availability zones requested is greater than the number
+    # of availability zones in the region, then the module should fail.
+    postcondition {
+      condition     = try(length(self.names) >= var.num_availability_zones, true)
+      error_message = "The number of availability zones requested is greater than the number of availability zones in the region."
+    }
+  }
+}
+
+
+# ------------------------------------------------------------
+
+# THE FOLLOWING SECTION IS USED TO CREATE THE VPC,
+
+# INTERNET GATEWAY (IGW), AND NAT GATEWAY
+
+# ------------------------------------------------------------
+
+locals {
+  # Gets a subset of the availability zones based on the number
+  # of availability zones requested.
+  availability_zones = (
+    var.num_availability_zones != null ?
+    slice(data.aws_availability_zones.current.names, 0, var.num_availability_zones) :
+    data.aws_availability_zones.current.names
+  )
+}
+
+
+# --------------------------------------------------------------------
+# CREATE THE VPC
+# --------------------------------------------------------------------
+
+# tfsec:ignore:aws-ec2-require-vpc-flow-logs-for-all-vpcs
+resource "aws_vpc" "this" {
+  cidr_block = "10.0.0.0/18"
+
+  #   ipv4_ipam_pool_id = "" # this should be set up
```

Comment:
This still needs to be implemented?

Reply:
I just removed it because I don't think anyone would use it.